prboom2/src/dsda/zipfile.c

```diff
@@ -0,0 +1,134 @@
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//	DSDA zipfile support using libzip
+//
+
+#include <stdio.h>
+#include <zip.h>
+
+#include "lprintf.h"
+#include "m_file.h"
+#include "z_zone.h"
+
+#include "dsda/utility.h"
+
+static char *dsda_JoinPath(const char *prefix, const char *suffix) {
+  int joined_path_length;
+  char *joined_path;
+
+  joined_path_length = snprintf(NULL, 0, "%s%s", prefix, suffix) + 1;
+  joined_path = (char *)Z_Malloc(joined_path_length);
+  if (joined_path == NULL) {
+    lprintf(LO_ERROR, "dsda_JoinPath: Failed to allocate joined path.\n");
+    return NULL;
+  }
+  snprintf(joined_path, joined_path_length, "%s%s", prefix, suffix);
+  return joined_path;
+}
+
+static void dsda_CreateSubDir(const char *directory, const char *subdirectory) {
+  char *full_path;
+  int result;
+
+  full_path = dsda_JoinPath(directory, subdirectory);
+  if (full_path == NULL) {
+    return;
+  }
+  result = M_MakeDir(full_path, false);
+  Z_Free(full_path);
+  if (result != 0) {
+    lprintf(LO_ERROR, "dsda_CreateSubDir: Failed to create subdirectory %s in %s.\n", subdirectory, directory);
+  }
+}
+
+static FILE *dsda_CreateFile(const char *directory, const char *filename) {
+  FILE *file_handle;
+  char *full_path;
+
+  full_path = dsda_JoinPath(directory, filename);
+  if (full_path == NULL) {
+    return NULL;
+  }
+  file_handle = M_OpenFile(full_path, "wb");
+  Z_Free(full_path);
+  return file_handle;
+}
+
+static void dsda_WriteContentToFile(zip_file_t *input_file, FILE *dest_file, zip_uint64_t data_size) {
+  char *buffer;
+
+  buffer = (char *)Z_Malloc(data_size);
+  if (buffer == NULL) {
+    lprintf(LO_ERROR, "dsda_WriteContentToFile: Could not allocate temporary buffer.\n");
+    return;
+  }
+  zip_fread(input_file, buffer, data_size);
+  fwrite(buffer, sizeof(char), data_size, dest_file);
+  Z_Free(buffer);
+}
+
+static void dsda_WriteZippedFilesToDest(zip_t *archive, const char *destination_directory) {
+  for (zip_int64_t i = 0; i < zip_get_num_entries(archive, ZIP_FL_UNCHANGED); i++) {
+    zip_file_t *zipped_file;
+    zip_stat_t stat;
+    FILE *dest_file_handle;
+    const char *file_name = zip_get_name(archive, i, ZIP_FL_UNCHANGED);
+
+    /* Intermediate directories show up as file */
+    if (dsda_HasFileExt(file_name, "/")) {
+      dsda_CreateSubDir(destination_directory, file_name);
+      continue;
+    }
+
+    zip_stat_index(archive, i, ZIP_FL_UNCHANGED, &stat);
+    if ((stat.valid & ZIP_STAT_SIZE) == 0) {
+      lprintf(LO_ERROR, "dsda_WriteZippedFilesToDest: Failed to read size of zipped file %s.\n", file_name);
+      return;
```

Comment:
If anything goes wrong in this process, we can call `I_Error` for issues instead of using `lprintf`, since it's a requirement that files coming through `-file` do exist. The `I_Error` function shows an error popup on windows and prints to the terminal otherwise, and does not return. Maybe at some point there will be a case for zip files where it's acceptable that the process fails, and we would need to add an option to trigger or not trigger `I_Error`, but I don't think we need to manage that right now.